File: 101.对称二叉树.py
#
# @lc app=leetcode.cn id=101 lang=python
#
# [101] 对称二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def dfs(root):
            if not root:
                res.append(None)
                return 
            dfs(root.left)
            res.append(root)
            dfs(root.right)
            return res
        
        def rdfs(root):
            if not root:
                rres.append(None) 
                return 
            rdfs(root.right)
            rres.append(root)
            rdfs(root.left)
            return rres
        res = []
        rres = []
        dfs(root)
        rdfs(root)

        return res == rres

        
        
# @lc code=end

File: 101.对称二叉树.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return helper(root, root);
    }
    //递归
    // bool helper(TreeNode* lnode, TreeNode* rnode){
    //     if(lnode == nullptr and rnode == nullptr)
    //         return true;
    //     else if (lnode == nullptr or rnode == nullptr)
    //         return false;
    //     else if(lnode->val == rnode->val)
    //         return(helper(lnode->left, rnode->right) and helper(lnode->right, rnode->left));
    //     else
    //     {
    //         return false;
    //     }
        
    // }
    //迭代
    bool helper(TreeNode* lnode, TreeNode* rnode){
        queue<TreeNode*> q;
        q.push(lnode);
        q.push(rnode);
        while(!q.empty()){
            lnode = q.front(); q.pop();
            rnode = q.front(); q.pop();
            if(!lnode and !rnode) continue;
            else if((!lnode || !rnode) or(lnode->val != rnode->val)) return false;
            
            q.push(lnode->left);
            q.push(rnode->right);
            q.push(lnode->right);
            q.push(rnode->left);

        }
        return true;
    }
    



};
// @lc code=end

